scanpy/external/pl.py

```diff
@@ -234,3 +239,443 @@ def sam(
+
+    Parameters
+    ----------
+    adata
+        Annotated data matrix.
+    no_bins
+        Number of bins for calculating marker density
+    smoothing_factor
+        Parameter controlling the degree of smoothing
+
+    Returns
+    -------
+    Updates `adata` with the following fields:
+
+    `weights_wishbone` : :class:`pandas.DataFrame` (`adata.uns`)
+        Computed gaussian weights for points at each location
+    `branch_point_bin` : 'int' (`adata.uns`)
+        Identifies the bin with the branch point. In case of no branching,
+        defaults to `no_bins`
+    `bins_wishbone` : :class:`numpy.ndarray` (`adata.uns`)
+        Computed bin locations and bin memberships
+    `branches_wishbone`: :class:`numpy.ndarray` (`adata.uns`)
+        In case of branching, returns a list of branches [2, 3].
+    """
+
+    # Compute bin locations and bin memberships
+    # Sort trajectory
+    trajectory = adata.obs['trajectory_wishbone'].sort_values()
+    bins = np.linspace(np.min(trajectory), np.max(trajectory), no_bins)
+
+    # Compute gaussian weights for points at each location
+    # Standard deviation estimated from Silverman's approximation
+    stdev = np.std(trajectory) * 1.34 * len(trajectory) ** (-1 / 5) * smoothing_factor
+    weights = np.exp(
+        -((np.tile(trajectory, [no_bins, 1]).T - bins) ** 2 / (2 * stdev ** 2))
+    ) * (1 / (2 * np.pi * stdev ** 2) ** 0.5)
+
+    # Adjust weights if data has branches
+    if 'branch_wishbone' in adata.obs.keys():
+        # Branch of the trunk
+        trunk = adata.obs['branch_wishbone'][trajectory.index[0]]
+        adata.uns['branches_wishbone'] = np.array([2, 3])
+        # Counts of branch cells in each bin
+        branch_counts = pd.DataFrame(np.zeros([len(bins) - 1, 3]), columns=[1, 2, 3])
+        for j in branch_counts.columns:
+            branch_counts[j] = pd.Series(
+                [
+                    pd.Series(
+                        adata.obs['branch_wishbone'][
+                            trajectory.index[
+                                (trajectory > bins[i - 1]) & (trajectory < bins[i])
+                            ]
+                        ]
+                        == j
+                    ).sum()
+                    for i in range(1, len(bins))
+                ]
+            )
+        # Frequencies
+        branch_counts = branch_counts.divide(branch_counts.sum(axis=1), axis=0)
+
+        # Identify the bin with the branch point by looking at the weights
+        weights = pd.DataFrame(weights, index=trajectory.index, columns=range(no_bins))
+        bp_bin = weights.columns[np.where(branch_counts[trunk] < 0.9)[0][0]] + 0
+        if bp_bin < 0:
+            bp_bin = 3
+
+    else:
+        bp_bin = no_bins
+
+    # return weight to adata
+    adata.uns['weights_wishbone'] = weights
+    adata.uns['branch_point_bin'] = bp_bin
+    adata.uns['bins_wishbone'] = bins
+
+
+@_doc_params(show_save_ax=doc_show_save_ax)
+def wishbone(
```

Comment:
This is really huge. Did you copy that from somewhere? If this is copied from the wishbone package, just call the one that’s there instead and write a little wrapper at most!

If you wrote all this code just for Scanpy: Thank you! Wouldn’t it be useful in the `wishbone` package though?

Reply:
Unlike other packages we have, Wishbone is actually built with 2 major class objects, one builds a data object to be used, as `adata` for Scanpy, and the second class has the core methods for wishbone, and plotting functions, that accept a Wishbone data object as input.
This plotting code is embedded and adapted from Wishbone, and requires the internal built in data object to be passed. To make it easier for Scanpy community, we decided, to adapt the code from the class in wishbone and make the proper changes such that it works with `AnnData` as input, instead. What you are suggesting to do, is actually wrapping and generating a new class object data using Wishbone and then passing this object to the internal plotting method. The internal plotting method will then generate plot, and additional data for plotting, useful for heatmaps, and stores them internally in the Wishbone data object. This is doable, but could create some confusion.

Reply:
Well, that’s easy to solve: You can just refactor the wishbone function to run with plain data instead, and make both your old plotting function and this one a wrapper around that one. Having two diverging but partially identical functions helps nobody.

Out of curiosity: Why do you use your own data object instead of AnnData?

Reply:
I think wishbone may be older than `AnnData` ;)

Reply:
Hi guys,

I work with Awni, and just wanted to drop a line here to have everyone on the same page.

None of the algorithms in Pe'er Lab was designed and built with Scanpy/AnnData in mind. For most, Scanpy/AnnData simply didn't exist back then :-) On top of that, both of us wasn't part of Pe'er Lab when these algorithms were developed (thus, we are not the original authors). Awni is simply trying to help integrate our algorithms into Scanpy. If you are asking why things were done this way or that way, I'm afraid that we have no good answer. It's like we have a USB 3.0, but you want USB Type-C. I really wish we could just go to Staples and buy an adapter :-)

Anyway, thanks for all your help!

Reply:
It not existing is the only good reason to not use AnnData :stuck_out_tongue_winking_eye: 

In all seriousness: I think it’s better for code health to use the same code. You are already adapting the original function, so what’s keeping you from doing that inside of the original package? The only references that function has to `self` are:

- `self.branch`
- `self.scdata.data.ix[trajectory.index, marker]`

So you could simply extract the whole method into a function that takes those two as parameters, and we can call that one from `sc.external.pl.wishbone_marker_trajectory` and from `Wishbone.plot_marker_trajectory`

That would mean 3 lines of code in scanpy and 3 additional lines of code in wishbone instead of 241 copied lines of code in scanpy.

---

All that being said: wishbone isn’t actively being maintained, so the chance of the code changing is low. So if I can trust you to come here and fix it if that’s not the case, I’m OK with you copying the dead code over here.

Reply:
I took your heed and made a simpler way to wrap the plotting function, without the need to change the original package, yet the new wrapper calls the original method for plotting, and maintains the same functionality. inside `sc.external.tl._wishbone`, I wrapped a data converter that generates a minimal dummy `wishbone` data object, that then can be passed easily to Wishbone's original plotting method in `sc.external.pl.wishbone_marker_trajectory`. It now looks much more cleaner, simpler, and most importantly `AnnData` and Scnapy friendly.

Reply:
One more thing, the plotting of marker trajectory generates data internally that is useful for further visualization. these data are pushed back to adata, as documented in the method.

Reply:
Perfect! that’s awesome, thank you.